### DISCUSSION TITLE
MOBIKE and DPD

### DISCUSSION THREAD

**Discussion by jfhren-stormshield** [strongswan/strongswan#1749] [General]
Hello, I stumbled upon a corner case with MOBIKE and DPD.

When a DPD delay is set in configuration and MOBIKE is supported by both peer, informational IKEv2 packets are exchanged with `NATD_S_IP` & `NATD_D_IP` notify payloads when there is no traffic for the set delay.

If there is always some traffic since the tunnel was mounted, no such informational IKEv2 packets were exchanged. If the initiator is behind some NAT and the mapping is changing, since no destination hash was stored in the IKE SA (through a call to `has_mapping_changed()` on it), the change is not detected. Since no informational IKEv2 packet with an `UPD_SA_ADDR` notify payload is sent to the responder, the old mapping is still in use leading to a loss of connectivity.

A solution might be to consider it a change if no destination hash was already set in the IKE SA.
Any thought on this matter ?

**Comment by tobiasbrunner** [comment]
> A solution might be to consider it a change if no destination hash was already set in the IKE SA.

What's not ideal is that this would always trigger a MOBIKE update after the first DPD. We'd basically get the same effect if we removed the exchange type check here:

https://github.com/strongswan/strongswan/blob/4c2747fbfc524801cf2d780f3160bca75196fd26/src/libcharon/sa/ikev2/tasks/ike_natd.c#L221-L229

As described in the comment, the hash stored there for IKE_SA_INIT will never match the one exchanged later due to the port switch. Not sure how much of an issue it would be to trigger such a possibly unnecessary `UPDATE_SA_ADDRESS`.

An alternative could be to forcibly send a DPD after establishing an IKE_SA with MOBIKE support.

**Reply by jfhren-stormshield** [reply]
Thanks for the reply.
The alternative would be something like changing the following condition
https://github.com/strongswan/strongswan/blob/4c2747fbfc524801cf2d780f3160bca75196fd26/src/libcharon/sa/ike_sa.c#L825-L832
to include a check on MOBIKE extension support & destination hash being empty.
Is that a valid approach ?

**Reply by tobiasbrunner** [reply]
Maybe, but depending on the DPD interval this increases the chances that the NAT mapping gets reset before a DPD could get sent. In particular, if a host doesn't have a DPD interval configured but  sends a DPD only in special circumstances e.g. triggered by `charon.keep_alive_dpd_margin` or `charon.check_current_path`. However, the change would actually immediately trigger a DPD on a responder that _has_ a DPD interval configured. That's because its task manger isn't busy, so a DPD task would immediately get queued and initiated. The resulting request would probably arrive at the client before the IKE_AUTH response, so it would drop it, i.e. this would require at least one retransmit.

I feel that triggering that potentially redundant MOBIKE address update with the first DPD might be the better approach. We could remove the check I pointed out above (so we'd have the hash from IKE_SA_INIT to compare to, which will always fail), or we just return `TRUE` here: https://github.com/strongswan/strongswan/blob/4c2747fbfc524801cf2d780f3160bca75196fd26/src/libcharon/sa/ike_sa.c#L1100-L1104

While this will cause an additional INFORMATIONAL exchange after the first DPD, it shouldn't be a problem. Actually, any MOBIKE-enabled DPD sent after an IKE_SA rekeying will also cause such an update as the NAT-D hashes include the SPIs (i.e. a hash cached before the rekeying will never match one exchanged afterwards). RFC 4555 even says:
    
    Note that this approach to detecting NAT mapping changes may cause an
    extra address update when the IKE_SA is rekeyed.  ...
    This unnecessary update is harmless, however.

**Reply by jfhren-stormshield** [reply]
Returning `TRUE` was the first change I tried and it worked in my case. Thank you again for the help.
